libcfnet/file_stream.c

```diff
@@ -0,0 +1,943 @@
+ * @param eof Is set to true if this was the last message in the transaction
+ * @return true on success, otherwise false
+ *
+ * @note RecvMessage fails if the communication is broken or if we received an
+ *       error from the remote host. In both cases, we should not try to flush
+ *       the stream.
+ */
+static bool RecvMessage(SSL *conn, char *msg, size_t *len, bool *eof)
+{
+    assert(conn != NULL);
+    assert(msg != NULL);
+    assert(len != NULL);
+    assert(eof != NULL);
+
+    /* TLSRecv() expects a buffer this size */
+    char recv_buffer[CF_BUFSIZE];
+
+    /* Receive header */
+    uint16_t header;
+    int ret = TLSRecv(conn, recv_buffer, HEADER_SIZE);
+    if (ret != HEADER_SIZE)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to receive message header during file stream (%d != %d)",
+            ret,
+            HEADER_SIZE);
+        return false;
+    }
+    memcpy(&header, recv_buffer, HEADER_SIZE);
+    header = ntohs(header);
+
+    /* Extract Error flag */
+    bool err = header & (1 << 0);
+
+    /* Extract End-of-File flag */
+    *eof = header & (1 << 1);
+
+    /* Extract message length */
+    *len = header >> 4;
+
+    /* Read payload */
+    if (*len > 0)
+    {
+        ret = TLSRecv(conn, recv_buffer, *len);
+        if (ret != *len)
+        {
+            Log(LOG_LEVEL_ERR,
+                "Failed to receive message payload during file stream (%d != %zu)",
+                ret,
+                *len);
+            return false;
+        }
+        memcpy(msg, recv_buffer, *len);
+
+        if (err)
+        {
+            /* If the error flag is set, then the payload contains an error
+             * message in the form of a NUL-byte terminated string. */
+            Log(LOG_LEVEL_ERR, "Remote file stream error: %s", msg);
+        }
+    }
+
+    return !err;
+}
+
+/**
+ * @brief Flush the file stream
+ *
+ * It's used to prevent the remote host from blocking while sending the
+ * remaining data after we have experienced an unexpected error and need to
+ * abort the file stream. Once the stream has been successfully flushed, the
+ * remote host will be ready to receive our error message.
+ *
+ * @param conn The SSL connection object
+ * @return true on success, otherwise false
+ */
+static bool FlushStream(SSL *conn)
+{
+    assert(conn != NULL);
+
+    char msg[MESSAGE_SIZE];
+    size_t len;
+    bool eof;
+    while (RecvMessage(conn, msg, &len, &eof))
+    {
+        if (eof)
+        {
+            return true;
+        }
+    }
+
+    Log(LOG_LEVEL_ERR, "Remote file stream error: %s", msg);
+    return false;
+}
+
+/**
+ * @brief Send an error message using the file stream protocol
+ *
+ * @param conn The SSL connection object
+ * @param flush Whether or not to flush the stream (see FlushStream())
+ * @param fmt The format string
+ * @param ... The format string arguments
+ * @return true on success, otherwise false
+ */
+static bool SendError(SSL *conn, bool flush, const char *fmt, ...)
```

Comment:
There's some `#pragma` to mark this as a printf-like function and let the compiler do all sorts of checks. Please add it.